### PR TITLE
refactor: migrate from `js` to `ts`

### PR DESCRIPTION
**Pull request by yoshinorin**

### CODE REVIEW

**Comment by SukkaW** [comment]
@yoshinorin Actually we have a better option.
Currently, @liby is working on porting hexo-util to Rust binding, to have better performance and speed (and first-party typescript typing supports). See https://github.com/liby/hexo-util-rs

**Comment by liby** [comment]
Hi, @yoshinorin If you're interested, you can take a look under the [`hexo-util-rs/tree/main/common`](https://github.com/liby/hexo-util-rs/tree/main/common) folder.

This folder holds all:
  1. Waiting to be RIIR
  2. Has been tried and is not suitable for RIIR
  
Welcome to your PR.

**Comment by yoshinorin** [comment]
Sounds interesting!
I will try learning Rust :)

**Comment by liby** [comment]
> I will try learning Rust :)

There are actually some other jobs that can be done, For example:
  1. Migrate the `.js` files in the *common* folder to `.ts`, if you think it's necessary.
  2. Improve the test.
  3. Improve type definition.
  4. ...

**Comment by stevenjoezhang** [comment]
We can merge this pull request first, and then start migrating to rust.

**Review by stevenjoezhang — DISMISSED** [review]

File: lib/color.ts [hunk 1]
```diff
@@ -176,21 +174,32 @@ const convertHue = (p, q, h) => {
...
 };
 
-const mixValue = (a, b, ratio) => a + ((b - a) * ratio);
+const mixValue = (a: number, b: number, ratio: number) => a + ((b - a) * ratio);
+
+interface RGBA {
+  r?: number;
+  g?: number;
+  b?: number;
+  a?: number;
```
Comment (yoshinorin) [inline lib/color.ts:190]: @stevenjoezhang 
Why `r, g, b, a` is optional?

File: lib/deep_merge.ts [hunk 2]
```diff
@@ -17,9 +15,8 @@ const arrayMerge = (target, source, options) => {
   return destination;
 };
 
-
 function deepMerge(target, source) {
```
Comment (yoshinorin) [inline lib/deep_merge.ts:18]:
```suggestion
function deepMerge<T>(target: Partial<T>, source: Partial<T>) {
```

**Comment by yoshinorin** [comment]
@stevenjoezhang 
Thank you ❤️ 
Let's release `v3.0.0` 🚀

**Pull request merged**